packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -106,87 +163,200 @@ export class BlobManager {
+                    this.sendBlobAttachOp(response.id);
+                }
+            } else if (entry.status === PendingBlobStatus.OfflinePendingUpload) {
+                entry.status = PendingBlobStatus.OfflinePendingOp;
+            }
+        } else {
+            // connected to storage but not ordering service?
+            this.logger.sendTelemetryEvent({ eventName: "BlobUploadSuccessWhileDisconnected" });
+            if (entry.status === PendingBlobStatus.OnlinePendingUpload) {
+                this.sendBlobAttachOp(response.id, localId);
+                entry.handleP.resolve(this.getBlobHandle(localId));
+            }
+            entry.status = PendingBlobStatus.OfflinePendingOp;
         }
+        return response;
+    }
 
-        // Note - server will de-dup blobs, so we might get existing blobId!
-        if (this.pendingBlobIds.has(response.id)) {
-            await this.pendingBlobIds.get(response.id)?.promise;
-        } else if (!this.blobIds.has(response.id)) {
-            this.pendingBlobIds.set(response.id, new Deferred<void>());
-
-            // send blob attach op and wait until we see it to return the handle
-            this.attachBlobCallback(response.id);
-            await this.pendingBlobIds.get(response.id)?.promise;
+    private async onUploadReject(localId: string, error) {
+        const entry = this.pendingBlobs.get(localId);
+        assert(!!entry, "Must have pending blob entry for blob which failed to upload");
+        if (!this.runtime.connected) {
+            entry.status = PendingBlobStatus.OfflinePendingUpload;
+            // since we are not connected, we will have a chance to add the storage ID when reSubmit() is called
+            this.sendBlobAttachOp(undefined, localId);
+            entry.handleP.resolve(this.getBlobHandle(localId));
+            // we are probably not connected to storage but start another upload request in case we are
+            const delay = this.retryThrottler.getDelay();
+            entry.uploadP = new Promise<void>((res) => setTimeout(res, delay))
+                .then(async () => this.uploadBlob(localId, entry.blob));
+            return entry.uploadP;
+        } else {
+            entry.handleP.reject(error);
+            throw error;
         }
+    }
 
-        return handle;
+    /**
+     * Resubmit a BlobAttach op. Used to add storage IDs to ops that were
+     * submitted to runtime while disconnected.
+     * @param metadata - op metadata containing storage and/or local IDs
+     */
+    public reSubmit(metadata: Record<string, unknown> | undefined) {
+        assert(!!metadata, "Resubmitted ops must have metadata");
+        const { blobId, localId }: { blobId?: string; localId?: string; } = metadata;
+        if (!blobId) {
+            assert(!!localId, "Submitted BlobAttach ops must have a blobId or localId");
+            // We submitted this op while offline. The blob should have been uploaded by now.
```

Comment:
I expect this isn't special to the offline case: we always generate the attach op after upload completes and therefore can assume that if resubmitting an attach op, the upload has previously completed.